platform/cxxsupport/mstd_span

```diff
@@ -0,0 +1,365 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MSTD_SPAN_
+#define MSTD_SPAN_
+
+#include <mstd_iterator>
+#include <mstd_type_traits>
+
+namespace mstd {
+
+inline constexpr size_t dynamic_extent = -1;
+
+template <typename, size_t = dynamic_extent> class span;
+
+namespace detail {
+
+template<typename ElementType, size_t Extent>
+class storage {
+public:
+    constexpr storage() noexcept = default;
+
+    constexpr storage(ElementType *ptr, size_t) noexcept :
+        _data(ptr)
+    {}
+
+    ElementType* _data;
+    static constexpr size_t _size = Extent;
+};
+
+template<typename ElementType>
+class storage<ElementType, dynamic_extent> {
+public:
+    constexpr storage() noexcept = default;
+
+    constexpr storage(ElementType *ptr, size_t size) noexcept :
+        _data(ptr), _size(size)
+    {}
+
+    ElementType* _data;
+    size_t _size;
+};
+
+template<typename>
+struct is_span: false_type {};
+
+template<typename T, size_t E>
+struct is_span<span<T, E>>: true_type {};
+
+template<typename>
+struct is_std_array: false_type {};
+
+template<typename T, size_t N>
+struct is_std_array<std::array<T, N>>: true_type {};
+
+template<typename, typename = void>
+struct has_size : std::false_type {};
+
+template<typename T>
+struct has_size<T, void_t<decltype(mstd::size(mstd::declval<T>()))>>:
+    std::true_type {};
+
+template<typename, typename = void>
+struct has_data : std::false_type {};
+
+template<typename T>
+struct has_data<T, void_t<decltype(mstd::data(mstd::declval<T>()))>>:
+    std::true_type {};
+
+template<typename T, typename U = mstd::remove_cvref_t<T>>
+struct is_container{
+    static constexpr bool value =
+        not is_span<U>::value && not is_std_array<U>::value &&
```

Comment:
`not` instead of `!` is rather unconventional, and not consistent with anything else in the codebase.

Reply:
You're right, there are some other style issues I need to fix too. Its hard to let go of my normal style for such things ;)